File: Easy/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {')': '(', ']': '[', '}': '{'}
        stack = []
        for char in s:
            if char in '([{':
                stack.append(char)
            else:
                if not stack:
                    return False
                last_bracket = stack.pop()
                if pairs[char] != last_bracket:
                    return False
        return not stack

if __name__ == '__main__': 
    sol = Solution()
    tests = ["()", "()[]{}", "(]", "([])", "([)]", '([)]']
    for s in tests:
        ans = sol.isValid(s)
        print(ans)

File: C++/Easy/20.cc
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        vector<int> stack;
        unordered_map<char, char> pairs = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        for (char c : s){
            if (pairs.find(c) != pairs.end()){
                if (stack.empty() || stack.back() != pairs[c]) return false;
                stack.pop_back();
            }
            else {
                stack.push_back(c);
            }
        }
        return stack.empty();               
    }
};